psi4/src/psi4/lib3index/df_helper.cc

```diff
@@ -1047,6 +1200,55 @@ std::string DF_Helper::compute_metric(double pow) {
     }
     return return_metfile(pow);
 }
+void DF_Helper::contract_metric_Qpq(std::string file, double* metp, double* Mp, double* Fp, const size_t tots) {
+    
+    std::string getf = std::get<0>(files_[file]);
+    std::string putf = std::get<1>(files_[file]);
+
+    size_t Q = std::get<0>(sizes_[getf]);
+    size_t l = std::get<1>(sizes_[getf]);
+    size_t r = std::get<2>(sizes_[getf]);
+    size_t count = 0;
+    size_t maxim = 0;
+
+    std::string op = "wb";
+
+    // determine blocking, block through l
+    std::vector<std::pair<size_t, size_t>> steps;
+    for (size_t i = 0; i < l; i++) {
+        count++;
+        if (tots < count * Q * r || i == l - 1) {
+            if (count == 1 && i != l - 1) {
+                std::stringstream error;
+                error << "DF_Helper:contract_metric: not enough memory.";
+                throw PSIEXCEPTION(error.str().c_str());
+            }
+            if (tots < count * Q * r) {
+                maxim = (maxim < count ? count - 1 : maxim);
+                steps.push_back(std::make_pair(i - count + 1, i - 1));
+                i--;
+            } else {
+                maxim = (maxim < count ? count : maxim);
+                steps.push_back(std::make_pair(i - count + 1, i));
+            }
+            count = 0;
+        }
+    }
```

Comment:
I trust that this works, but I think it would be a bit difficult to understand the logic. Can you add a comment on how this works or see if there is a simpler way to figure out the blocking?

Reply:
Good point, this is getting a bit blurred as I try to make it general with all possibilities.  I will see what I can do.